timesketch/api/v1/resources/event.py

```diff
@@ -1208,3 +1248,167 @@ def post(self, sketch_id):
+                "Cannot untag more than {0:d} events in a single "
+                "request".format(self.MAX_EVENTS_TO_TAG),
+            )
+
+        tags = form.get("tags", [])
+        if not isinstance(tags, list):
+            abort(HTTP_STATUS_CODE_BAD_REQUEST, "Tags need to be a list")
+
+        if len(tags) > self.MAX_TAGS_PER_REQUEST:
+            abort(
+                HTTP_STATUS_CODE_BAD_REQUEST,
+                "Cannot untag more than {0:d} tags in a single "
+                "request".format(self.MAX_TAGS_PER_REQUEST),
+            )
+
+        datastore = self.datastore
+
+        for _event in events:
+            # every event entry can have a dedicated searchindex_id or searchindex_name
+            searchindex_id = _event.get("searchindex_id", None)
+            searchindex_name = _event.get("_index", None)
+
+            if not (searchindex_id or searchindex_name):
+                abort(
+                    HTTP_STATUS_CODE_NOT_FOUND,
+                    "No search index information supplied.",
+                )
+
+            searchindex = None
+            # in both cases we are flexible, no matter what was supplied
+            if searchindex_name:
+                searchindex = SearchIndex.query.filter_by(
+                    index_name=searchindex_name
+                ).first()
+            elif searchindex_id:
+                searchindex = SearchIndex.query.get(searchindex_id)
+
+            if not searchindex:
+                abort(
+                    HTTP_STATUS_CODE_BAD_REQUEST,
+                    "Unable to find the Search index.",
+                )
+
+            if searchindex.get_status.status == "deleted":
+                abort(
+                    HTTP_STATUS_CODE_BAD_REQUEST,
+                    "Unable to query event on a closed search index.",
+                )
+
+            result = self.datastore.get_event(searchindex.index_name, _event.get("_id"))
+            if not result:
+                logger.warning(
+                    "Unable to find event %s in index %s to untag",
+                    _event.get("_id"),
+                    searchindex.index_name,
+                )
+                abort(
+                    HTTP_STATUS_CODE_NOT_FOUND,
+                    "Unable to find event {0:s} in index {1:s} to untag".format(
+                        _event.get("_id"), searchindex.index_name
+                    ),
+                )
+
+            existing_tags = result.get("_source").get("tag", [])
+
+            new_tags = list(set(existing_tags) - set(tags))
+
+            if existing_tags == new_tags:
+                continue
+
+            # write the new tags to the datastore
+            datastore.import_event(
+                index_name=searchindex.index_name,
+                event_id=_event.get("_id"),
+                event={"tag": new_tags},
+                flush_interval=datastore.DEFAULT_FLUSH_INTERVAL,
+            )
+            datastore.flush_queued_events()
```

Comment:
You want to do this outside of the loop. Otherwise you will flush the queue on every iteration and lose the benefit of the bulk change.

Reply:
done